### PR TITLE
*dm, util-linux, openssh: adjust PATH for usrmerge

### PR DESCRIPTION
**Pull request by Chocimier**
/bin is still added to PATH by /etc/profile, but after /usr/bin.
This makes `which` return real file instead of symlink.

### CODE REVIEW

**Comment by Piraty** [comment]
It's the other way around you mean? `which ls` gives `/bin/ls` (which is the symlink) but should be `/usr/bin/ls` (which is the real file).
I stumbled on this before, as `xbps-query -o $(which ls)` fails because of this.

**Comment by Chocimier** [comment]
> It's the other way around you mean? |which ls| gives |/bin/ls| (which is the symlink) but should be |/usr/bin/ls| (which is the real file).

Yes, the effect of that change is that /usr/bin/ls is returned.

On tty at least, because lxdm sets PATH independently. Would we need to patch each display manager?

**Comment by ericonr** [comment]
From `./configure --help`:

```
 --enable-usrdir-path    use only /usr paths in PATH env. variable
                          (recommended on systems with /bin -> /usr/bin
                          symlinks)
```

I think we should add this change, but also fix `/etc/profile` to put `/usr/bin` in front of everything.

**Comment by Chocimier** [comment]
/etc/profile already have /usr/bin in front of /bin. Problem here is that /etc/profile appends to PATH passed to shell from it's parent, and every parent (atty, ssh, different graphical login managers) passes something different, usually not fitting usrmerge.

Unless there is vaild usecase for preserving parent's path, it could be reset by /etc/profile instead of patching all possible parents.

**Comment by Chocimier** [comment]
#6233 states it is for WSL, optimizing for that doesn't convince me.
Is it still relevant with WSL2?

**Comment by ericonr** [comment]
It seems @am11's https://github.com/am11/VoidWSL project is still active, at least...

**Comment by am11** [comment]
Yup, we are actively providing https://git.io/voidwsl and https://git.io/voidmuslwsl from GitHub releases. It is mainly the Void Linux official rootf image with a [slim WSL C binding](https://github.com/yuk7/wsldl/blob/main/main.c).

**Comment by ericonr** [comment]
I'm a bit confused now, isn't `which` provided by the `which` package, which is completely separate from `util-linux`?

**Comment by Chocimier** [comment]
`which` looks for executable in directories of `$PATH`, in order, as passed to it by shell.
Shell build `$PATH` as specified in /etc/profile, that is by appending to `$PATH` passed from parent process.
Parent process may be agetty from `util-linux`, ssh, graphical login manager like lxdm.
Order of /etc/profile is right, but all directories there are read by `which` after directories from parent, that for some parents are symlinks, leading to reporting symlink directories rather than real ones.

Given there is some usage of not reset PATH in profile, I would like to sync defaults of login managers.

**Comment by Piraty** [comment]
Let's resurrect this

**Comment by Chocimier** [comment]
@Piraty, rebased. Would you like to patch services you use?

**Comment by 0x5c** [comment]
I use KDE from SDDM, and the default path is `/usr/local/bin:/usr/bin:/bin:/usr/local/sbin:/usr/sbin:/sbin`.
It seems correct already but different than the path specified in this PR.

**Comment by Chocimier** [comment]
As @leahneukirchen already started doing this in 3b9cc95ac4157290d16dc46fe087b6072367a991, I will patch few more *dm and merge.

**Comment by leahneukirchen** [comment]
The PR title should be adjusted.

File: srcpkgs/openssh/template [hunk 1]
```diff
@@ -1,13 +1,14 @@
...
 version=9.3p1
-revision=2
+revision=3
 build_style=gnu-configure
 configure_args="--datadir=/usr/share/openssh
  --sysconfdir=/etc/ssh --without-selinux --with-privsep-user=nobody
  --with-mantype=doc --without-rpath --with-xauth=/usr/bin/xauth
  --disable-strip --with-privsep-path=/var/chroot/ssh
- --with-default-path=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
+ --with-default-path=/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/sbin:/sbin:/bin
```
Comment (leahneukirchen) [inline srcpkgs/openssh/template:10]: Why put /usr/bin before /usr/sbin, yet /sbin before /bin? IMO sbin always should go first, as on classic Unix.

File: srcpkgs/util-linux/patches/default-PATH.patch [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+@@ -19,20 +19,12 @@
+ /* DEFPATHs from <paths.h> don't include /usr/local */
+ #undef _PATH_DEFPATH
+ 
+-#ifdef USE_USRDIR_PATHS_ONLY
+-# define _PATH_DEFPATH	        "/usr/local/bin:/usr/bin"
+-#else
+-# define _PATH_DEFPATH	        "/usr/local/bin:/bin:/usr/bin"
+-#endif
++# define _PATH_DEFPATH	        "/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/sbin:/sbin:/bin"
```
Comment (leahneukirchen) [inline srcpkgs/util-linux/patches/default-PATH.patch:20]: Same here.

File: srcpkgs/util-linux/patches/default-PATH.patch [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ 
+ #undef _PATH_DEFPATH_ROOT
+ 
+-#ifdef USE_USRDIR_PATHS_ONLY
+-# define _PATH_DEFPATH_ROOT	"/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin"
+-#else
+-# define _PATH_DEFPATH_ROOT	"/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin"
+-#endif
+-
++# define _PATH_DEFPATH_ROOT	"/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/sbin:/sbin:/bin"
```
Comment (leahneukirchen) [inline srcpkgs/util-linux/patches/default-PATH.patch:30]: And here.

File: srcpkgs/sddm/patches/0.19.0--default-path.patch [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+diff --git a/src/common/Configuration.h b/src/common/Configuration.h
+index cf44a62..4c01c4a 100644
+--- a/src/common/Configuration.h
++++ b/src/common/Configuration.h
+@@ -82,7 +82,7 @@ namespace SDDM {
+         );
+ 
+         Section(Users,
+-            Entry(DefaultPath,         QString,     _S("/usr/local/bin:/usr/bin:/bin"),         _S("Default $PATH for logged in users"));
++            Entry(DefaultPath,         QString,     _S("/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/sbin:/sbin:/bin"), _S("Default $PATH for logged in users"));
```
Comment (leahneukirchen) [inline srcpkgs/sddm/patches/0.19.0--default-path.patch:16]: And here.

**Review by leahneukirchen — CHANGES_REQUESTED** [review]

**Comment by Chocimier** [comment]
/usr/bin goes before /usr/sbin to make `command -v`, zsh `=` expansion and so on return real path of command, so it is more useful, e.g. can be passed to `xbps-query -o`. It is the order given in /etc/profile as well. Using real path is main reason to introduce changes here, as it worked until #6233..

**Comment by leahneukirchen** [comment]
This makes sense, thanks for pointing it out.

**Review by leahneukirchen — APPROVED** [review]

**Pull request merged**